### PR TITLE
Improved package : petsc

### PR DESCRIPTION
**Pull request by alalazo**
##### Modifications : 
- [x] added variants for `mpi`, `shared`, `double` (floating point precision)
- [x] added variants for a bunch of supported dependencies

##### Note :
After `do_install` the build will possibly fail with an error that is unrelated to the package modifications: 
```
Traceback (most recent call last):
  File "/home/culpo/github/spack/bin/spack", line 176, in <module>
    main()
  File "/home/culpo/github/spack/bin/spack", line 154, in main
    return_val = command(parser, args)
  File "/home/culpo/github/spack/lib/spack/spack/cmd/install.py", line 81, in install
    fake=args.fake)
  File "/home/culpo/github/spack/lib/spack/spack/package.py", line 946, in do_install
    spack.hooks.post_install(self)
  File "/home/culpo/github/spack/lib/spack/spack/hooks/__init__.py", line 69, in __call__
    hook(pkg)
  File "/home/culpo/github/spack/lib/spack/spack/hooks/sbang.py", line 75, in post_install
    if shebang_too_long(path):
  File "/home/culpo/github/spack/lib/spack/spack/hooks/sbang.py", line 39, in shebang_too_long
    with open(path, 'r') as script:
IOError: [Errno 21] Is a directory: '/home/culpo/github/spack/opt/spack/x86_E5v2_IntelIB/gcc-4.4.7/petsc-3.6.3-oaidkcmi2x6fggx2dclhu6lh63elebvi/bin/julia'
```
I didn't solve that yet and I am not sure if we want it to be part of this PR. I assume it may be related to #497.

### CODE REVIEW

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 1]
```diff
@@ -1,40 +1,86 @@
...
-    depends_on("hypre")
-    depends_on("parmetis")
-    depends_on("metis")
-    depends_on("hdf5+mpi")
-    depends_on("mpi")
+    # Other dependencies
+    depends_on('boost', when='+boost')
+    depends_on('metis', when='+metis')
+
+    depends_on('hdf5~cxx~unsupported+mpi', when='+hdf5+mpi')
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/petsc/package.py]: does petsc conflict with `hdf5+cxx` or does it just not need the C++ bindings?  Likewise for unsupported -- would `depends_on('hdf5+mpi')` be enough or does that cause issues?

Comment (eschnett) [inline var/spack/repos/builtin/packages/petsc/package.py]: I'm building PETSc with `hdf5+cxx+unsupported`, and it's working fine. I think this needs changing.

Comment (tgamblin) [inline var/spack/repos/builtin/packages/petsc/package.py]: I merged it -- want to just submit another PR?

Comment (alalazo) [inline var/spack/repos/builtin/packages/petsc/package.py]: @eschnett For me `hdf5+cxx+unsupported` works with anything but Intel. That's why I used the supported version.

Comment (tgamblin) [inline var/spack/repos/builtin/packages/petsc/package.py]: Would it make sense to put `depends_on('hdf5+mpi', when='+mpi')` here and make the `unsupported` variant in `hdf5` default to `False`?  If it breaks with intel we probably don't want it to be enabled by default.

Comment (alalazo) [inline var/spack/repos/builtin/packages/petsc/package.py]: @tgamblin I think Intel is a strange beast. After the e-mail exchange with @luigicalori on the spack group I retried #451 and I noticed a regression on OpenMPI (I'll try to investigate that tomorrow). My point is : I am not sure that what I see with Intel on my platform is what everybody sees on their platforms. On the other hand I don't think `petsc` needs the c++ bindings (even though if you want `petsc` and `hdf5+cxx` installed on your system then you end up with two installations of `hdf5`)

Comment (eschnett) [inline var/spack/repos/builtin/packages/petsc/package.py]: `unsupported` in HDF5's configurations doesn't mean that it enables dodgy code. It only enables additional configuration options, or rather combinations of configuration options. Thus, if something doesn't work with Intel, something else needs to be disabled, and `unsupported` can remain enabled.

Here, it seems that the Intel compiler can't handle HDF5's C++ interface.

Which version of the Intel compiler is this? What is the error message? What system is this?

If we disable something, we should keep a record of why this was done.

Comment (tgamblin) [inline var/spack/repos/builtin/packages/petsc/package.py]: @alalazo: you actually don't (necessarily).  `depends_on('hdf5+mpi')` is not the same as `depends_on('hdf5+mpi~cxx')`.  The first can link against an `hdf5` that satisfies `depends_on('hdf5+mpi+cxx')`, if it's available.  The second cannot.

Comment (eschnett) [inline var/spack/repos/builtin/packages/petsc/package.py]: The current PETSc dependency means that no one can use PETSc and C++-HDF5 at the same time.

Comment (alalazo) [inline var/spack/repos/builtin/packages/petsc/package.py]: @eschnett I think my position on having `unsupported` as a default is quite clear so I won't argue on that again :smile:

@tgamblin It was just a choice of convenience : petsc doesn't need `+cxx` and having the spec restricted permits to always use the default without appending something like:
```
spack install petsc%intel ^ hdf5~cxx~unsupported
```
If you think it is better having it less restricted it's fine with me.

Comment (tgamblin) [inline var/spack/repos/builtin/packages/petsc/package.py]: I think less restricted is good.  I think we can work on the inconvenience aspect when PR #120 is merged, which I need to get back to :smile:.  Now to resolve this unsupported thing...
